Fix author_roles defn. Add links

hornc opened the pull request:
Both of these are "Kind: embeddable" from the descriptions at https://openlibrary.org/type/author_role and https://openlibrary.org/type/link  as such they should have `type` set with the format `"type": { "key": "/type/<type>"}`. Updating the json schema for both to reflect this.

Unfortunately in the current work data for author_roles, there is a pretty even mix of `type` styles:
`"authors": [{"type": "/type/author_role"`   => **8,280,196**
vs
`"authors": [{"type": {"key": "/type/author_role"}`  => **7,743,184**

'links` in works are fully consistent, all use the "type: key:" style, I used that as the model for how an OL embeddable type should be.

hornc commented:
This PR is based off the old master without my updated tests ans is just a change to the json schema. Not sure why the tests are failing, I think the old expectations were wrong...

cdrini commented on olclient/schemata/work.schema.json:
```diff
@@ -59,10 +66,18 @@
-        "author": { "$ref": "#/definitions/author" }
+        "author": { "$ref": "#/definitions/author" },
+        "role":   { "type": "string" },
+        "as":     { "type": "string" }
```
Ooooh! I've never seen `as` before, do you have an example? Is it for pseudonyms?

hornc replied:
@cdrini Unfortunately I don't think `as` is used at all in the data, but it appears to be part of the intended design as documented https://openlibrary.org/type/author_role

`role` is only used a couple of times in the data, and currently the UI has nothing to display it, but I would love to add something to show editor, translator, 'introduction by' or whatever else. 

My interpretation  of `as` is for aliases or pseudonyms too, which would be great to capture and display.

The `author_role` property makes sense only with these extra fields though, otherwise a work would just contain a list of `authors`. I had been wondering why works had this extra level of data to specify author. We should take advantage of it!

There is a separate conversation to have about where `author_roles` should live. Currently `author_roles` is only on works. Editions have an optional plain `authors` list, but the UI ignores edition authors and only displays work authors.

I would suggest it is appropriate to have `author_roles` at an edition level too since translators only apply to the translated edition, editors can change over time, and other contributors can be involved in producing a particular edition. An edition should display a combined work + edition `author_roles` list of unique author+role combinations. Again, probably a bigger discussion to be had later! :)

cdrini commented:
Lgtm. We should definitely move `definitions` into a separate file in the future though, so that we don't have duplicate definitions for `link` ( https://spacetelescope.github.io/understanding-json-schema/structuring.html ).

mekarpeles commented:
@hornc it looks like the error which is being thrown is related to a discrepancy between the `author_key` field:

    'author_key': u'__new__' v. 'author_key': u'/authors/OL7292805A' in test_openlibrary.py L70

hornc commented:
@mek I have rebased on the recently merged mocked tests  which pass. That old test seemed to be wrong, expecting `__new__` for the existing author  `u'Karl Schwarzer'` = `u'/authors/OL7292805A'` was a bug in the code I fixed a while ago in another PR: https://github.com/internetarchive/openlibrary-client/pull/30/commits/d9f889a71979f680dfd6026ede4e5056af103c1f

This PR passes now! :)

hornc commented:
@cdrini re. shared definitions, I thought the definitions had to be shared via absolute URLs, and it would be nice to have the schema eventually hosted and official at https://openlibrary.org  , but it looks like relative paths work too. I'll need to check the python tools support that correctly.

mekarpeles commented:
Great! Lgtm
